src/trace.zig

```diff
@@ -0,0 +1,338 @@
+pub const Event = enum {
+    replica_commit,
+
+    compact_blip_read,
+    compact_blip_merge,
+    compact_blip_write,
+    compact_manifest,
+    compact_mutable,
+
+    const stack_count = std.meta.fields(Event).len;
+
+    // Stack is a u32 since it must be losslessly encoded as a JSON integer.
+    fn stack(event: *const Event) u32 {
+        return @intFromEnum(event.*);
+    }
+};
+
+pub const Tracer = struct {
+    replica_index: u8,
+    options: Options,
+
+    events_enabled: [Event.stack_count]bool = [_]bool{false} ** Event.stack_count,
+    time_start: std.time.Instant,
+
+    pub const Options = struct {
+        /// The tracer still validates start/stop state even when writer=null.
+        writer: ?std.io.AnyWriter = null,
+    };
+
+    pub fn init(allocator: std.mem.Allocator, replica_index: u8, options: Options) !Tracer {
+        _ = allocator;
+
+        if (options.writer) |writer| {
+            try writer.writeAll("[\n");
+        }
+
+        return .{
+            .replica_index = replica_index,
+            .options = options,
+            .time_start = std.time.Instant.now() catch @panic("std.time.Instant.now() unsupported"),
+        };
+    }
+
+    pub fn deinit(tracer: *Tracer, allocator: std.mem.Allocator) void {
+        _ = allocator;
+
+        tracer.* = undefined;
+    }
+
+    pub fn start(tracer: *Tracer, event: Event, data: anytype) void {
+        comptime assert(@typeInfo(@TypeOf(data)) == .Struct);
+
+        const stack = event.stack();
+        assert(!tracer.events_enabled[stack]);
+        tracer.events_enabled[stack] = true;
+
+        log.debug("{}: {s}: start:{}", .{
+            tracer.replica_index,
+            @tagName(event),
+            struct_format(data, .dense),
+        });
+
+        const writer = tracer.options.writer orelse return;
+        const time_now = std.time.Instant.now() catch unreachable;
+        const time_elapsed_ns = time_now.since(tracer.time_start);
+        const time_elapsed_us = @divFloor(time_elapsed_ns, std.time.ns_per_us);
+
+        // String tid's would be much more useful.
+        // They are supported by both Chrome and Perfetto, but rejected by Spall.
+        writer.print("{{" ++
+            "\"pid\":{[process_id]}," ++
+            "\"tid\":{[thread_id]}," ++
+            "\"cat\":\"{[category]s}\"," ++
+            "\"ph\":\"{[event]c}\"," ++
+            "\"ts\":{[timestamp]}," ++
+            "\"name\":\"{[name]s}{[data]}\"," ++
+            "\"args\":{[args]}" ++
+            "}},\n", .{
+            .process_id = tracer.replica_index,
+            .thread_id = event.stack(),
+            .category = @tagName(event),
+            .event = 'B',
+            .timestamp = time_elapsed_us,
+            .name = @tagName(event),
+            .data = struct_format(data, .sparse),
+            .args = std.json.Formatter(@TypeOf(data)){ .value = data, .options = .{} },
+        }) catch unreachable;
```

Comment:
That ` .args = std.json.Formatter(@TypeOf(data)){ .value = data, .options = .{} },` does't result in too large JSON output. 

That is to prevent outright mistakes like accidentally including the entire block into a message, but also to prevent boiling frog type of situation, where, for some event, we just add more and more fields, until it becomes hard to analyse. 

I guess, one background thought here is that perhaps we could format the even into a fixed-sized internal buffer first, and then dump it to write as one single slice? Potentially this could be a bit faster I think.

Reply:
Ah that makes sense, thanks!

> we could format the even into a fixed-sized internal buffer first, and then dump it to write as one single slice? Potentially this could be a bit faster I think.

I think this is what `BufferedWriter` does, which I need to add anyhow.

Reply:
I was goign to say that we can avoid BufferedWriter and do a single `fmt.bufPrint`, but it uses the writer internally anyway: 

```
pub fn bufPrint(buf: []u8, comptime fmt: []const u8, args: anytype) BufPrintError![]u8 {
    var fbs = std.io.fixedBufferStream(buf);
    format(fbs.writer().any(), fmt, args) catch |err| switch (err) {
        error.NoSpaceLeft => return error.NoSpaceLeft,
        else => unreachable,
    };
    return fbs.getWritten();
}
```

I wonder if at some point we should do `fmt` which also tracks the worst-case size of the output, so that we statically check that the size of buffer is big enough? Sounds like a bigger adventure though!

Reply:
> I guess, one background thought here is that perhaps we could format the even into a fixed-sized internal buffer first, and then dump it to write as one single slice? Potentially this could be a bit faster I think.

One thought with https://github.com/tigerbeetle/tigerbeetle/pull/2266/commits/edcc13f93b2ff9b7625e94ebf2c35861579b3282 is to make sure that we're not affected by the `write()` to the file blocking. Probably good enough for now to use a path on tmpfs as the target.

Reply:
> Probably good enough for now to use a path on tmpfs as the target.

That, and tracing is behind `--experimental`. Long-term we should probably use io_uring for logging (which will include tracing).